EssentialApp/EssentialAppTests/FeedCommentsUIIntegrationTests.swift

```diff
@@ -0,0 +1,257 @@
+	}
+	
+	func test_loadingFeedCommentsIndicator_isVisibleWhileLoadingFeedComments() {
+		let (sut, loader) = makeSUT()
+		
+		sut.loadViewIfNeeded()
+		XCTAssertTrue(sut.isShowingLoadingIndicator, "Expected loading indicator once view is loaded")
+		
+		loader.completeCommentsLoading(at: 0)
+		XCTAssertFalse(sut.isShowingLoadingIndicator, "Expected no loading indicator once loading completes successfully")
+		
+		sut.simulateUserInitiatedFeedCommentsReload()
+		XCTAssertTrue(sut.isShowingLoadingIndicator, "Expected loading indicator once user initiates a reload")
+		
+		loader.completeCommentsLoadingWithError(at: 1)
+		XCTAssertFalse(sut.isShowingLoadingIndicator, "Expected no loading indicator once user initiated loading completes with error")
+	}
+	
+	func test_loadFeedCommentsCompletion_rendersSuccessfullyLoadedFeedComments() {
+		let comment0 = makeComment(date: Date().adding(days: -1))
+		let comment1 = makeComment(date: Date().adding(hours: -1))
+		let comment2 = makeComment(date: Date().adding(mins: -3))
+		let (sut, loader) = makeSUT()
+		
+		sut.loadViewIfNeeded()
+		assertThat(sut, isRendering: [])
+		
+		loader.completeCommentsLoading(with: [comment0], at: 0)
+		assertThat(sut, isRendering: [comment0])
+		
+		sut.simulateUserInitiatedFeedCommentsReload()
+		loader.completeCommentsLoading(with: [comment0, comment1, comment2], at: 1)
+		assertThat(sut, isRendering: [comment0, comment1, comment2])
+	}
+	
+	func test_loadFeedCommentsCompletion_rendersSuccessfullyLoadedEmptyCommentsAfterNonEmptyComments() {
+		let comment0 = makeComment()
+		let comment1 = makeComment()
+		let (sut, loader) = makeSUT()
+		
+		sut.loadViewIfNeeded()
+		loader.completeCommentsLoading(with: [comment0, comment1], at: 0)
+		assertThat(sut, isRendering: [comment0, comment1])
+		
+		sut.simulateUserInitiatedFeedCommentsReload()
+		loader.completeCommentsLoading(with: [], at: 1)
+		assertThat(sut, isRendering: [])
+	}
+	
+	func test_loadFeedCommentsCompletion_doesNotAlterCurrentRenderingStateOnError() {
+		let comment0 = makeComment()
+		let (sut, loader) = makeSUT()
+		
+		sut.loadViewIfNeeded()
+		loader.completeCommentsLoading(with: [comment0], at: 0)
+		assertThat(sut, isRendering: [comment0])
+		
+		sut.simulateUserInitiatedFeedCommentsReload()
+		loader.completeCommentsLoadingWithError(at: 1)
+		assertThat(sut, isRendering: [comment0])
+	}
+	
+	func test_loadFeedCommentsCompletion_rendersErrorMessageOnErrorUntilNextReload() {
+		let (sut, loader) = makeSUT()
+		
+		sut.loadViewIfNeeded()
+		XCTAssertEqual(sut.errorMessage, nil)
+		
+		loader.completeCommentsLoadingWithError(at: 0)
+		XCTAssertEqual(sut.errorMessage, localized("FEED_COMMENTS_VIEW_CONNECTION_ERROR"))
+		
+		sut.simulateUserInitiatedFeedCommentsReload()
+		XCTAssertEqual(sut.errorMessage, nil)
+	}
+	
+	func test_errorViewIsHidden_whenUserTapsOnIt() {
+		let (sut, loader) = makeSUT()
+		sut.loadViewIfNeeded()
+		loader.completeCommentsLoadingWithError(at: 0)
+		
+		sut.errorView.simulateTap()
```

Comment:
Missing pre-condition to increase coverage.

```suggestion
		loader.completeCommentsLoadingWithError(at: 0)
		XCTAssertEqual(sut.errorMessage, localized("FEED_COMMENTS_VIEW_CONNECTION_ERROR"))
		
		sut.errorView.simulateTap()
```